chain/chain/src/store.rs

```diff
@@ -1745,111 +1780,243 @@ impl<'a> ChainStoreUpdate<'a> {
+            self.chain_store.chunks.cache_remove(&chunk_header.hash.clone().into());
+            // 3b. Delete chunk extras (ColChunkExtra)
+            store_update.delete(ColChunkExtra, chunk_header.hash.as_ref());
+            self.chain_store.chunk_extras.cache_remove(&chunk_header.hash.clone().into());
+            // 3c. Delete partial_chunks (ColPartialChunks)
+            store_update.delete(ColPartialChunks, chunk_header.hash.as_ref());
+            self.chain_store.partial_chunks.cache_remove(&chunk_header.hash.clone().into());
+            // 3d. Delete invalid chunks (ColInvalidChunks)
+            store_update.delete(ColInvalidChunks, chunk_header.hash.as_ref());
+            self.chain_store.invalid_chunks.cache_remove(&chunk_header.hash.clone().into());
+        }
+
+        // 4. Delete block_hash-indexed data
+        // 4a. Delete block (ColBlock) if not genesis
+        if height > self.get_genesis_height() {
+            store_update.delete(ColBlock, block_hash.as_ref());
+            self.chain_store.blocks.cache_remove(&block_hash.clone().into());
+        }
+        // 4b. Delete block header (ColBlockHeader) - don't do because header sync needs headers
+        // 4c. Delete block extras (ColBlockExtra)
+        store_update.delete(ColBlockExtra, block_hash.as_ref());
+        self.chain_store.block_extras.cache_remove(&block_hash.clone().into());
+        // 4d. Delete from next_block_hashes (ColNextBlockHashes)
+        store_update.delete(ColNextBlockHashes, block_hash.as_ref());
+        self.chain_store.next_block_hashes.cache_remove(&block_hash.clone().into());
+        // 4e. Delete from my_last_approvals (ColMyLastApprovalsPerChain)
+        store_update.delete(ColMyLastApprovalsPerChain, block_hash.as_ref());
+        self.chain_store.my_last_approvals.cache_remove(&block_hash.clone().into());
+        // 4f. Delete from ColChallengedBlocks
+        store_update.delete(ColChallengedBlocks, block_hash.as_ref());
+        // 4g. Delete from ColBlocksToCatchup
+        store_update.delete(ColBlocksToCatchup, block_hash.as_ref());
+        // 4h. Delete from KV state changes
+        let storage_key = KeyForStateChanges::get_prefix(&block_hash);
+        // 4h1. We should collect all the keys which key prefix equals to `block_hash`
+        let changes_iter = self
+            .chain_store
+            .store()
+            .iter_prefix_ser::<Vec<u8>>(ColStateChanges, storage_key.as_ref());
+        // 4h2. Remove from ColStateChanges all found KV changes
+        for change in changes_iter {
+            let (key, _) = change?;
+            store_update.delete(ColStateChanges, &key);
+        }
+
+        if is_fork {
+            // 5. Update block_hash_per_height - for forks only
+            let mut epoch_to_hashes = self
+                .chain_store
+                .block_hash_per_height
+                .cache_get(&index_to_bytes(height))
+                .expect("height exists")
```

Comment:
why does this have to exist in cache? It is possible that this height has already been pushed out for example. Also it is better to use `cache_remove` here to avoid cloning.

Reply:
I want to remove `block_hash` from `block_hash_per_height` collection. Could you suggest me the best way to do that?
For now, I found the following solution:
```
            // 5. Update block_hash_per_height - for forks only
            let epoch_to_hashes_ref =
                self.get_all_block_hashes_by_height(height).expect("current height exists");
            let mut hashes = epoch_to_hashes_ref
                .get(&block.header.inner_lite.epoch_id)
                .cloned()
                .expect("current epoch id should exist");
            hashes.remove(&block_hash);
            let mut epoch_to_hashes = epoch_to_hashes_ref.clone();
            epoch_to_hashes.insert(block.header.inner_lite.epoch_id, hashes);

            store_update.set_ser(ColBlockPerHeight, &index_to_bytes(height), &epoch_to_hashes)?;
            self.chain_store.block_hash_per_height.cache_remove(&index_to_bytes(height));
```

Reply:
looks fine to me.